test/unsafekzg/kzgsrs.go

```diff
@@ -0,0 +1,393 @@
+	fft_bls24315 "github.com/consensys/gnark-crypto/ecc/bls24-315/fr/fft"
+	fft_bls24317 "github.com/consensys/gnark-crypto/ecc/bls24-317/fr/fft"
+	fft_bn254 "github.com/consensys/gnark-crypto/ecc/bn254/fr/fft"
+	fft_bw6633 "github.com/consensys/gnark-crypto/ecc/bw6-633/fr/fft"
+	fft_bw6761 "github.com/consensys/gnark-crypto/ecc/bw6-761/fr/fft"
+
+	fr_bls12377 "github.com/consensys/gnark-crypto/ecc/bls12-377/fr"
+	fr_bls12381 "github.com/consensys/gnark-crypto/ecc/bls12-381/fr"
+	fr_bls24315 "github.com/consensys/gnark-crypto/ecc/bls24-315/fr"
+	fr_bls24317 "github.com/consensys/gnark-crypto/ecc/bls24-317/fr"
+	fr_bn254 "github.com/consensys/gnark-crypto/ecc/bn254/fr"
+	fr_bw6633 "github.com/consensys/gnark-crypto/ecc/bw6-633/fr"
+	fr_bw6761 "github.com/consensys/gnark-crypto/ecc/bw6-761/fr"
+
+	bls12377 "github.com/consensys/gnark-crypto/ecc/bls12-377"
+	bls12381 "github.com/consensys/gnark-crypto/ecc/bls12-381"
+	bls24315 "github.com/consensys/gnark-crypto/ecc/bls24-315"
+	bls24317 "github.com/consensys/gnark-crypto/ecc/bls24-317"
+	"github.com/consensys/gnark-crypto/ecc/bn254"
+	bw6633 "github.com/consensys/gnark-crypto/ecc/bw6-633"
+	bw6761 "github.com/consensys/gnark-crypto/ecc/bw6-761"
+)
+
+var (
+	cache           map[uint64]cacheEntry
+	memLock, fsLock sync.RWMutex
+)
+
+func init() {
+	cache = make(map[uint64]cacheEntry)
+}
+
+// NewSRS returns a pair of kzg.SRS; one in canonical form, the other in lagrange form.
+// Default options use a memory cache, see Option for more details & options.
+func NewSRS(ccs constraint.ConstraintSystem, opts ...Option) (canonical kzg.SRS, lagrange kzg.SRS, err error) {
+	nbConstraints := ccs.GetNbConstraints()
+	sizeSystem := nbConstraints + ccs.GetNbPublicVariables()
+
+	sizeLagrange := ecc.NextPowerOfTwo(uint64(sizeSystem))
+	sizeCanonical := sizeLagrange + 3
+
+	curveID := utils.FieldToCurve(ccs.Field())
+
+	cfg, err := options(opts...)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	memLock.RLock()
+	key := cacheKey(curveID, sizeCanonical)
+	entry, ok := cache[key]
+	memLock.RUnlock()
+	if ok {
+		return entry.canonical, entry.lagrange, nil
+	}
+
+	if cfg.fsCache {
+		fsLock.RLock()
+		entry, ok := fsRead(key, cfg.cacheDir)
+		fsLock.RUnlock()
+		if ok {
+			canonical, lagrange = entry.canonical, entry.lagrange
+			memLock.Lock()
+			cache[key] = cacheEntry{canonical, lagrange}
+			memLock.Unlock()
+			return
+		}
+	}
+
+	// not in cache, generate
+	canonical, lagrange, err = newSRS(curveID, sizeCanonical)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// cache it
+	memLock.Lock()
+	cache[key] = cacheEntry{canonical, lagrange}
+	memLock.Unlock()
+
+	if cfg.fsCache {
+		fsLock.Lock()
+		fsWrite(key, cfg.cacheDir, canonical, lagrange)
+		fsLock.Unlock()
+	}
+
+	return canonical, lagrange, nil
+}
+
+type cacheEntry struct {
+	canonical kzg.SRS
+	lagrange  kzg.SRS
+}
+
+func cacheKey(curveID ecc.ID, size uint64) uint64 {
+	return uint64(curveID)<<59 | size
```

Comment:
I would rather use readable key names. When we cache to disk then it may be beneficial to have intelligible names for sharing etc. For example `kzg_srs-{{ CurveID }}-size_{{ size }}`.